runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillSink.java

```diff
@@ -214,8 +214,27 @@ public long add(WindowedValue<T> data) throws IOException {
               .setData(value)
               .setMetadata(metadata);
       keyedOutput.addMessages(builder.build());
+
+      long offsetSize = 0;
+      if (context.offsetBasedDeduplicationEnabled()) {
+        byte[] rawId = context.getCurrentRecordId();
+        if (rawId.length == 0) {
+          throw new RuntimeException("Unexpected empty record ID while offset-based deduplication enabled.");
+        }
+        id = ByteString.copyFrom(rawId);
```

Comment:
should we prefer this id or the id from ValueWithRecordId if both are set?
or we could make it not valid to have both set or expect them to be the same if both are set.

Reply:
Updated to not expect ID from ValueWithRecordId.